### PR TITLE
pin megatron-fsdp and add backwards tests

### PR DESCRIPTION
**Pull request by pstjohn**
Fix models/esm2 tests by pinning megatron-fsdp and adding additional backwards tests

## Summary by CodeRabbit

- Tests
  - Expanded FP8 tests to include backward passes.
  - Added THD test suite covering collator parity, inference loss checks, and backpropagation.
  - Conditional handling for newer GPUs (SM120) and attention-fusion toggling in tests.
  - More reliable test setup by reloading the TE backend; improved MXFP8 availability guard.

- Chores
  - Pinned megatron-fsdp to 0.1.0rc0 for reproducible installs.

- Breaking Changes
  - Removed the TensorFlow experimental compilation option from the MLM data-collator constructor.

### CODE REVIEW

File: bionemo-recipes/models/esm2/tests/test_thd.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+    input_data_bshd = {k: v.to("cuda") for k, v in input_data_bshd.items()}
+    input_data_thd = {k: v.to("cuda") if isinstance(v, torch.Tensor) else v for k, v in input_data_thd.items()}
+
+    bshd_outputs = model_bshd(**input_data_bshd)
+    thd_outputs = model_thd(**input_data_thd)
+
+    torch.testing.assert_close(bshd_outputs.loss, thd_outputs.loss)
+
+    # bshd_logits = bshd_outputs.logits[input_data_bshd["attention_mask"].to(bool)]
```
Comment (jomitchellnv) [inline bionemo-recipes/models/esm2/tests/test_thd.py:97]: delete?

Comment (pstjohn) [reply bionemo-recipes/models/esm2/tests/test_thd.py:97]: no we need that line for when we look into BIONEMO-2801
